onmt/translate/BeamSearchAdvancer.lua

```diff
@@ -0,0 +1,74 @@
+--[[ Class for specifying how to advance one step
+
+
+           ______________
+          / (update beam)\
+         /                \
+      scores, states       \     .................    nil, nil
+        /                   \    /                 \     /
+    step()                   \ step()               \  step()
+       /                      \ /                    \ /
+  nil, nil---------------tokens, states........tokens, states
+
+      |                        |                      |
+
+    t = 1--------------------t = 2..............t = max seq len
+
+ ==================================================================
+
+Speicifies how to go one step forward.
+--]]
+local BeamSearchAdvancer = torch.class('BeamSearchAdvancer')
+
+--[[Step function. BeamSearcher will take care of updating beams and
+  transforming states, such that this function only need to take a batched input
+  and produce a batched output. All tensors (either parameters or returns,
+  including tensors inside tables) must have the same first dimension `batchSize`.
+  Note that since BeamSearcher will remove completed hypotheses, the `batchSize`
+  does not remain between steps, hence this function cannot assume a static one.
+
+Parameters:
+
+  * `tokens` - a flat tensor of size `batchSize`, predictions from the previous step. For the first step, it would be nil.
+  * `states` - the states from the previous step. For the first step, it would be nil.
+
+Returns:
+
+  * `scores` - a 2D tensor of size `(batchSize, vocabSize)`, scores for the current step. For the last step, return nil.
+  * `states` - the states for the current step. For the last step, return nil.
+
+]]
+function BeamSearchAdvancer:step(tokens, states)
+  local scores, statesOut, t
+  if tokens == nil then
+    -- The first step, batchSize 3, vocabSize 5, hiddenSize 7 and 8
```

Comment:
Is this a test implementation? Shouldn't be in the main code. Instead put in test.lua.